pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractTopic.java

```diff
@@ -382,17 +382,25 @@ public String getReplicatorPrefix() {
             switch (producer.getAccessMode()) {
             case Shared:
                 if (hasExclusiveProducer || !waitingExclusiveProducers.isEmpty()) {
-                    return FutureUtil.failedFuture(new ProducerBusyException(
-                            "Topic has an existing exclusive producer: " + producers.keys().nextElement()));
+                    StringBuilder errorMsg = new StringBuilder();
+                    errorMsg.append("Topic has an existing exclusive producer");
+                    if (producers.keys().hasMoreElements()) {
+                        errorMsg.append(": " + producers.keys().nextElement());
```

Comment:
I'm not sure this avoids the race altogether since there's no atomicity between checking `hasMoreElements()` and `nextElement()`. I think we could get an iterator and use that to check, since it will be guaranteed to be consistent.